[GCC13 only] array-bounds error during compilation of testFWCoreUtilitiesCatch2

In GCC13 IB, the following errors were emitted when building `testFWCoreUtilitiesCatch2`:

```
>> Building binary testFWCoreUtilitiesCatch2
<...>
In member function '__dt_base ',
    inlined from '__dt_base ' at src/FWCore/Utilities/interface/clone_ptr.h:11:10,
    inlined from '__dt_base ' at src/FWCore/Utilities/test/test_catch2_clone_ptr.cc:26:10,
    inlined from 'destroy' at /data/cmsbld/jenkins/workspace/build-any-ib/w/el9_amd64_gcc13/external/gcc/13.2.0-1b0a3367d04f48f01ad3ccf40e55475c/include/c++/13.2.0/bits/new_allocator.h:194:13,
    inlined from 'destroy' at /data/cmsbld/jenkins/workspace/build-any-ib/w/el9_amd64_gcc13/external/gcc/13.2.0-1b0a3367d04f48f01ad3ccf40e55475c/include/c++/13.2.0/bits/alloc_traits.h:557:15,
    inlined from '_M_realloc_insert' at /data/cmsbld/jenkins/workspace/build-any-ib/w/el9_amd64_gcc13/external/gcc/13.2.0-1b0a3367d04f48f01ad3ccf40e55475c/include/c++/13.2.0/bits/vector.tcc:507:28,
    inlined from 'push_back' at /data/cmsbld/jenkins/workspace/build-any-ib/w/el9_amd64_gcc13/external/gcc/13.2.0-1b0a3367d04f48f01ad3ccf40e55475c/include/c++/13.2.0/bits/stl_vector.h:1289:21,
    inlined from '____C_A_T_C_H____T_E_S_T____0' at src/FWCore/Utilities/test/test_catch2_clone_ptr.cc:84:19:
  /data/cmsbld/jenkins/workspace/build-any-ib/w/el9_amd64_gcc13/external/gcc/13.2.0-1b0a3367d04f48f01ad3ccf40e55475c/include/c++/13.2.0/bits/unique_ptr.h:403:19: error: array subscript 0 is outside array bounds of 'struct B[0:]' [-Werror=array-bounds=]
   403 |         if (__ptr != nullptr)
      |                   ^
In function '____C_A_T_C_H____T_E_S_T____0':
lto1: note: source object is likely at address zero
In member function '__dt_base ',
    inlined from 'destroy' at /data/cmsbld/jenkins/workspace/build-any-ib/w/el9_amd64_gcc13/external/gcc/13.2.0-1b0a3367d04f48f01ad3ccf40e55475c/include/c++/13.2.0/bits/new_allocator.h:194:13,
    inlined from 'destroy' at /data/cmsbld/jenkins/workspace/build-any-ib/w/el9_amd64_gcc13/external/gcc/13.2.0-1b0a3367d04f48f01ad3ccf40e55475c/include/c++/13.2.0/bits/alloc_traits.h:557:15,
    inlined from '_M_realloc_insert' at /data/cmsbld/jenkins/workspace/build-any-ib/w/el9_amd64_gcc13/external/gcc/13.2.0-1b0a3367d04f48f01ad3ccf40e55475c/include/c++/13.2.0/bits/vector.tcc:507:28,
    inlined from 'push_back' at /data/cmsbld/jenkins/workspace/build-any-ib/w/el9_amd64_gcc13/external/gcc/13.2.0-1b0a3367d04f48f01ad3ccf40e55475c/include/c++/13.2.0/bits/stl_vector.h:1289:21,
    inlined from '____C_A_T_C_H____T_E_S_T____0' at src/FWCore/Utilities/test/test_catch2_clone_ptr.cc:84:19:
  src/FWCore/Utilities/test/test_catch2_clone_ptr.cc:26:10: error: array subscript 0 is outside array bounds of 'struct B[0:]' [-Werror=array-bounds=]
    26 |   struct B {
      |          ^
In function '____C_A_T_C_H____T_E_S_T____0':
lto1: note: source object is likely at address zero
```

[full log](https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/el9_amd64_gcc13/CMSSW_14_1_X_2024-06-26-1200/FWCore/Utilities)

--- a/FWCore/Utilities/interface/clone_ptr.h
+++ b/FWCore/Utilities/interface/clone_ptr.h
@@ -8,12 +8,13 @@ namespace extstd {
   /* modify unique_ptr behaviour adding a "cloning" copy-constructor and assignment-operator
    */
   template <typename T>
-  struct clone_ptr : public std::unique_ptr<T> {
+  struct clone_ptr {
+    clone_ptr() noexcept : ptr_() {}
     template <typename... Args>
-    explicit clone_ptr(Args&&... args) noexcept : std::unique_ptr<T>(std::forward<Args>(args)...) {}
+    explicit clone_ptr(Args&&... args) noexcept : ptr_(std::forward<Args>(args)...) {}
 
-    clone_ptr(clone_ptr const& rh) : std::unique_ptr<T>(rh ? rh->clone() : nullptr) {}
-    clone_ptr(clone_ptr&& rh) noexcept : std::unique_ptr<T>(std::move(rh)) {}
+    clone_ptr(clone_ptr const& rh) : ptr_(rh ? rh->clone() : nullptr) {}
+    clone_ptr(clone_ptr&& rh) noexcept : ptr_(std::move(rh.ptr_)) {}
 
     clone_ptr& operator=(clone_ptr const& rh) {
       if (&rh != this)
@@ -22,14 +23,16 @@ namespace extstd {
     }
     clone_ptr& operator=(clone_ptr&& rh) noexcept {
       if (&rh != this)
-        std::unique_ptr<T>::operator=(std::move(rh));
+        ptr_ = std::move(rh.ptr_);
       return *this;
     }
 
+    operator bool() const { return static_cast<bool>(ptr_); }
+
     template <typename U>
-    clone_ptr(clone_ptr<U> const& rh) : std::unique_ptr<T>(rh ? rh->clone() : nullptr) {}
+    clone_ptr(clone_ptr<U> const& rh) : ptr_(rh ? rh->clone() : nullptr) {}
     template <typename U>
-    clone_ptr(clone_ptr<U>&& rh) noexcept : std::unique_ptr<T>(std::move(rh)) {}
+    clone_ptr(clone_ptr<U>&& rh) noexcept : ptr_(std::move(rh.ptr_)) {}
 
     template <typename U>
     clone_ptr& operator=(clone_ptr<U> const& rh) {
@@ -40,9 +43,20 @@ namespace extstd {
     template <typename U>
     clone_ptr& operator=(clone_ptr<U>&& rh) noexcept {
       if (&rh != this)
-        std::unique_ptr<T>::operator=(std::move(rh));
+        ptr_ = std::move(rh.ptr_);
       return *this;
     }
+
+    T* operator->() const { return ptr_.get(); }
+
+    T& operator*() const { return *ptr_; }
+
+    T* get() const { return ptr_.get(); }
+
+    void reset(T* iValue) { ptr_.reset(iValue); }
+
+  private:
+    std::unique_ptr<T> ptr_;
   };
 
 }  // namespace extstd

--- a/FWCore/Utilities/test/test_catch2_clone_ptr.cc
+++ b/FWCore/Utilities/test/test_catch2_clone_ptr.cc
@@ -73,6 +73,7 @@ TEST_CASE("Test clone_ptr", "[clone_ptr]") {
   SECTION("in vector") {
     A::Guard ag;
     std::vector<B> vb(1);
+    vb.reserve(2);
     B b;
     b.a.reset(new A(-2));
     B c;

